domain/aggregator/scan_state_aggregator.go

```diff
@@ -0,0 +1,253 @@
+	agg.workingDirectoryScanStates[FolderProductKey{Product: product.ProductInfrastructureAsCode, FolderPath: folderPath}] = &ScanState{Status: NotStarted}
+}
+
+// AddNewFolder adds new folder to the state aggregator map with initial NOT_STARTED state
+func (agg *ScanStateAggregator) AddNewFolder(folderPath string) {
+	agg.mu.Lock()
+	agg.initForAllProducts(folderPath)
+	agg.mu.Unlock()
+	agg.scanStateChangeEmitter.Emit(agg)
+}
+
+// SetScanState changes the Status field of the existing state (or creates it if it doesn't exist).
+func (agg *ScanStateAggregator) SetScanState(folderPath string, p product.Product, isReferenceScan bool, newState ScanState) {
+	agg.mu.Lock()
+	agg.setScanState(folderPath, p, isReferenceScan, newState)
+	agg.mu.Unlock()
+
+	agg.scanStateChangeEmitter.Emit(agg)
+}
+
+func (agg *ScanStateAggregator) setScanState(folderPath string, p product.Product, isReferenceScan bool, newState ScanState) {
+	logger := agg.c.Logger().With().Str("method", "SetScanState").Logger()
+
+	key := FolderProductKey{FolderPath: folderPath, Product: p}
+	var st *ScanState
+	var exists bool
+	if isReferenceScan {
+		st, exists = agg.referenceScanStates[key]
+	} else {
+		st, exists = agg.workingDirectoryScanStates[key]
+	}
+
+	if !exists {
+		logger.Error().Msgf("Scan State for folder path%s and product %s doesn't exist in state aggregator", folderPath, p.ToProductNamesString())
+		return
+	}
+
+	st.Status = newState.Status
+	st.Err = newState.Err
+}
+
+func (agg *ScanStateAggregator) SetScanDone(folderPath string, p product.Product, isReferenceScan bool, scanErr error) {
+	agg.mu.Lock()
+
+	state := ScanState{}
+	if scanErr != nil {
+		state.Status = Error
+		state.Err = scanErr
+	} else {
+		state.Status = Success
+	}
+
+	agg.setScanState(folderPath, p, isReferenceScan, state)
+	agg.mu.Unlock()
+	agg.scanStateChangeEmitter.Emit(agg)
+}
+
+func (agg *ScanStateAggregator) SetScanInProgress(folderPath string, p product.Product, isReferenceScan bool) {
+	agg.mu.Lock()
+
+	state := ScanState{
+		Status: InProgress,
+	}
+
+	agg.setScanState(folderPath, p, isReferenceScan, state)
+	agg.mu.Unlock()
+
+	agg.scanStateChangeEmitter.Emit(agg)
+}
+
+func (agg *ScanStateAggregator) AllScansStarted(isReference bool) bool {
+	agg.mu.RLock()
+	defer agg.mu.RUnlock()
+
+	var stateMap ScanStateMap
+	if isReference {
+		stateMap = agg.referenceScanStates
+	} else {
+		stateMap = agg.workingDirectoryScanStates
+	}
+
+	for _, st := range stateMap {
+		if st.Status == NotStarted {
+			return false
+		}
+	}
+	return true
+}
+
+func (agg *ScanStateAggregator) AnyScanInProgress(isReference bool) bool {
+	agg.mu.RLock()
+	defer agg.mu.RUnlock()
+
+	var stateMap ScanStateMap
+	if isReference {
+		stateMap = agg.referenceScanStates
+	} else {
+		stateMap = agg.workingDirectoryScanStates
+	}
+
+	for _, st := range stateMap {
+		if st.Status == InProgress {
+			return true
+		}
+	}
```

Comment:
This looks very similar to other funcs - I think it can be extracted and the value to check can be passed into it.

Reply:
Also the return value, to maximise reuse ;)

Reply:
Could even be a predicate that is passed in as a function to return with the given early return value in the loop